### PR TITLE
Support float32 type in datastream

### PR DESCRIPTION
**Pull request by arawind**
This is a patch of https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/1375.

Successfully ran the dataflow template:

MySQL source, stored to GCS via datastream:
Screenshot from 2024-07-10 01-12-34

Spanner destination from datastream:
Screenshot from 2024-07-10 01-12-52

```
export PROJECT=span-cloud-testing
export IMAGE_NAME=datastream-to-spanner
export BUCKET_NAME=gs://shubhamswe-test
export TARGET_GCR_IMAGE=gcr.io/${PROJECT}/templates/${IMAGE_NAME}
export BASE_CONTAINER_IMAGE=gcr.io/dataflow-templates-base/java11-template-launcher-base
export BASE_CONTAINER_IMAGE_VERSION=latest
export APP_ROOT=/template/${IMAGE_NAME}
export DATAFLOW_JAVA_COMMAND_SPEC=${APP_ROOT}/resources/${IMAGE_NAME}-command-spec.json
#export SPANNER_HOST="https://staging-wrenchworks.sandbox.googleapis.com/"
export INSTANCE_ID="arawind-test"
export DATABASE_ID="float32-test"
export GCS_LOCATION="gs://shubhamswe-test/float32-mysql-out"
export STREAM_NAME="projects/span-cloud-testing/locations/us-central1/streams/arawind-mysql-datastream"
export TEMPLATE_IMAGE_SPEC="gs://shubhamswe-test/images/2024_07_09_01/flex/Cloud_Datastream_to_Spanner"
export JOB_NAME="${IMAGE_NAME}-`date +%Y%m%d-%H%M%S-%N`"

gcloud dataflow flex-template run ${JOB_NAME} \
        --project=${PROJECT} --region=us-central1 \
        --template-file-gcs-location=${TEMPLATE_IMAGE_SPEC} \
        --parameters instanceId=${INSTANCE_ID},databaseId=${DATABASE_ID},inputFilePattern=${GCS_LOCATION},streamName=${STREAM_NAME}
```

### CODE REVIEW

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventTypeConvertorTest.java [hunk 1]
```diff
@@ -205,6 +205,81 @@ public void cannotConvertNonExistentRequiredFieldToLong() throws Exception {
...
+    assertEquals(
+        ChangeEventTypeConvertor.toFloat(ce, "field1", /* requiredField= */ true),
+        new Long(123457));
+  }
+
+  @Test(expected = ChangeEventConvertorException.class)
+  public void cannotConvertNonExistentRequiredFieldToFloat() throws Exception {
+    JSONObject changeEvent = new JSONObject();
+    JsonNode ce = getJsonNode(changeEvent.toString());
+    assertEquals(
```
Comment (shreyakhajanchi) [inline v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventTypeConvertorTest.java:278]: nit: Maybe better to just do  `ChangeEventTypeConvertor.toFloat(ce, "field1", /* requiredField= */ true)` for exception cases instead of assert, its a bit confusing

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java [hunk 2]
```diff
@@ -81,6 +81,9 @@ static Ddl getTestDdl() {
             .column("int64_field")
             .int64()
             .endColumn()
+            .column("float32_field")
```
Comment (shreyakhajanchi) [inline v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java:84]: Shouldn't there be changes in v2/datastream-to-spanner/src/main/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertor.java?

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/ddl/InformationSchemaScannerTest.java [hunk 3]
```diff
@@ -355,16 +367,18 @@ public void testScanGSQLDdl() {
     Ddl ddl = informationSchemaScanner.scan();
     String expectedDdl =
         "CREATE TABLE `singer` (\n"
-            + "\t`singerId`                              STRING(50) NOT NULL,\n"
+            + "\t`singerId`                              INT64 NOT NULL,\n"
```
Comment (shreyakhajanchi) [inline v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/ddl/InformationSchemaScannerTest.java:370]: why are we modifying the datatype of other columns?

Comment (arawind) [reply v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/ddl/InformationSchemaScannerTest.java:370]: This is just to increase the coverage. The types were only testing STRING, added tests for INT64. Why INT64? Felt like SingerId would be an INT64 :)

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java (hunk 2 above)
Comment (arawind) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java:84]: No, that file only has transformations for primary keys. FLOAT32 is not a primary keyable type, which is also why we do not add it to the list of primary keys below

[the remark this answers is not in the document]
Comment (arawind) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java:84]: float_colum already exists in that test. I've modified spanner-schema.sql to correct the mapping from FLOAT64 to FLOAT32

File: v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventTypeConvertorTest.java (hunk 1 above)
Comment (arawind) [reply v2/spanner-common/src/test/java/com/google/cloud/teleport/v2/spanner/migrations/convertors/ChangeEventTypeConvertorTest.java:278]: Done

File: v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java (hunk 2 above)
Comment (shreyakhajanchi) [reply v2/datastream-to-spanner/src/test/java/com/google/cloud/teleport/v2/templates/datastream/ChangeEventConvertorTest.java:84]: yeah makes sense, i didn't notice that the file code had been refactored

**Comment by manitgupta** [comment]
Can you explain what am I looking at in the screenshots above? I think the change in value of `amount1` is expected? Can you link the docs here for me to read?

Is this change "normal" and "expected" - i.e will customers be okay with this change?

1.23 in MySQL becomes 1.230000002 in Spanner.

**Review by manitgupta — COMMENTED** [review]
The code changes themselves look good, thanks for the rigor on updating the ITs as well.

I want to better understand the implications on precision of the data while migrating a source float value to Float32 with respect to how the behaviour changes from the current mapping of Float64.

**Comment by arawind** [comment]
Thanks @manitgupta!

> Can you explain what am I looking at in the screenshots above? I think the change in value of `amount1` is expected? Can you link the docs here for me to read?
> 
> Is this change "normal" and "expected" - i.e will customers be okay with this change?
> 
> 1.23 in MySQL becomes 1.230000002 in Spanner.

TL;DR: The issue is just in the Cloud Console UI, the actual value of 1.23 is still preserved. 
The underlying reason is because we're up-casting the `float` value to `double` while on the Cloud Spanner API, as that's the best we can do with the limited types that the API provides.
Client libraries do the conversions float-to-double while sending, and double-to-float while receiving, and these conversions are lossless.
Now, since Cloud Console uses a JS client, and since JS does not support a float32 type, we're doing our best to convert the float64 value in the API to float32 in JS, but it is just an approximation to the actual down casting, and we end up with these values on the UI.

The actual values are preserved across migrations, import / export etc.

Design doc is available internally at go/cloud-spanner-float32.

**Review by shreyakhajanchi — APPROVED** [review]
LGTM

**Review by manitgupta — APPROVED** [review]
Thank you @arawind!
